### ISSUE TITLE
RadDataform DatePicker returns a timestamp when started with value.

### ISSUE THREAD

**Issue by marcelomiranda90** [telerik/nativescript-ui-feedback#942]
### Did you verify this is a real problem by searching the [NativeScript Forum](https://discourse.nativescript.org/)?
Yes

### Tell us about the problem

RadDataForm DatePicker when started with a value yyyy-MM-dd returns a timestamp and not a string.

Only after the change to any other date, the value returned is a string, as expected.

### Which platform(s) does your issue occur on?
Android 7.1

### Please provide the following version numbers that your issue occurs with:
- **Progress NativeScript UI version:** (the version of nativescript-telerik-ui/-pro in the `package.json`)
- **CLI: 5.0.1
- **Cross-platform modules:** 5.0.3
- **Runtime(s):** 5.0.0
- nativescript-ui-dataform: 3.9.0

### Is there code involved? If so, please share the minimal amount of code needed to recreate the problem.

Download the example and upgrade to RadDataForm for version 3.9.0, launch the application, click the button and watch the console.

https://play.nativescript.org/?template=play-ng&id=WvsZX0

**Comment by pbgc** [comment]
I have exactly this problem!
In a ui dataform if I initialize a datepicker field with "2018-11-30" and then change it to "2018-12-01" .. the returned value is the string "2018-12-01".
If I don't initialize (I don't want to .. I want the field to be initially empty) and then choose "2018-12-01" the returned value is: 1543687642999
This only happend after migrating to TNS 5 / nativescript-ui-dataform: 3.9.0, ... The same project with TNS 4.2 / nativescript-ui-dataform: 3.7.0 works as intended.

**Comment by tgpetrov** [comment]
Hi @pbgc 

There have been a few fixes in nativescript-ui-dataform, since version 3.7.0, which you can see listed [here](https://github.com/telerik/nativescript-ui-feedback/blob/master/releases/dataform.md#390-2018-november-20). Some of them may be affecting the way you have been using the control:
- The DatePicker/TimePicker editors now support both type Date and type String [871](https://github.com/telerik/nativescript-ui-feedback/issues/871)
- The Validation is applied to keys instead of to values [797](https://github.com/telerik/nativescript-ui-feedback/issues/797)

Now that the Preview app is updated and contains nativescript-ui-dataform@3.9.0, I used the [playground link](https://play.nativescript.org/?template=play-ng&id=WvsZX0) from the previous post without downloading the app and in the console the string value was printed as expected. Can you confirm?

Also I was not able to reproduce the issue with the SegmentedEditor ignoring values. Can you try to reproduce it in the Playground and post the link here so we can assist you further.

As to the value of the EntityProperty, exposing the label as a value was a bug, however we can consider exposing the label now through another property, if requested.

**Comment by pbgc** [comment]
Hi @tgpetrov 
Thank you for your answer.

> There have been a few fixes in nativescript-ui-dataform, since version 3.7.0, which you can see listed [here](https://github.com/telerik/nativescript-ui-feedback/blob/master/releases/dataform.md#390-2018-november-20). Some of them may be affecting the way you have been using the control:
> 
>     * The DatePicker/TimePicker editors now support both type Date and type String [871](https://github.com/telerik/nativescript-ui-feedback/issues/871)
> 
>     * The Validation is applied to keys instead of to values [797](https://github.com/telerik/nativescript-ui-feedback/issues/797)
> 

It really seems that those changes affected my code...

> Now that the Preview app is updated and contains nativescript-ui-dataform@3.9.0, I used the [playground link](https://play.nativescript.org/?template=play-ng&id=WvsZX0) from the previous post without downloading the app and in the console the string value was printed as expected. Can you confirm?

I'm going to take a look ... but at least in my app if I don't initialize the value on the datepicker (the value in the being null so the datepicker don't show an initial date) then the only way of getting the string I got before (with 3.7.0) is doing:
 ```
let movDateProperty = dataform.getPropertyByName("mov_date")
let date = new Date(Number(movDateProperty.value))
let movDateStr = `${data.getFullYear()}-${data.getMonth()}-${data.getDate()}`
```

> Also I was not able to reproduce the issue with the SegmentedEditor ignoring values. Can you try to reproduce it in the Playground and post the link here so we can assist you further.

I will try to do it .. because in my app for example in the following code, SegmentEditor is completely ignoring initial values .... 

```
<df:EntityProperty name="tipo" displayName="Tipo Assistência Técnica:" index="2" valuesProvider="{{ tipoValues }}">
  <df:EntityProperty.editor>
      <df:PropertyEditor type="SegmentedEditor">
          <df:PropertyEditor.propertyEditorStyle>
              <df:PropertyEditorStyle labelFontStyle="Bold" labelTextColor="#12673D"/>
           </df:PropertyEditor.propertyEditorStyle>
       </df:PropertyEditor>
   </df:EntityProperty.editor>
</df:EntityProperty>
```
Just changing from "SegmentEditor" to "Picker" in the code above and it works.
 
> As to the value of the EntityProperty, exposing the label as a value was a bug, however we can consider exposing the label now through another property, if requested.

PLEASE! I have a big app that does that assumption (it uses the values/ids on the observable to save data on the database and the string/Entity Property value to show it after on a ListView ... and If I don't have that property I will have to change all the app.

**Comment by pbgc** [comment]
@tgpetrov 

> Also I was not able to reproduce the issue with the SegmentedEditor ignoring values. Can you try to > reproduce it in the Playground and post the link here so we can assist you further.

Here it is. Before Segmented Editor would show the option 1 selected and now ignores the initial value !!!

https://play.nativescript.org/?template=play-js&id=0NqLMI&v=6

> As to the value of the EntityProperty, exposing the label as a value was a bug, however we can consider exposing the label now through another property, if requested.

How can I request more than already did in the last post ?

**Comment by pbgc** [comment]
@tsonevn @tgpetrov any news about this? I've sent the requested info (nativescript playground link). And what about the possibility of exposing the label through another property ?

**Comment by pbgc** [comment]
@tgpetrov did you saw the playground link I've sent? (you asked for it) any news about the other things? (exposing the label)

**Comment by tgpetrov** [comment]
Hi @pbgc 

I apologise for the delayed reply.

Thank you for the provided playground link, I managed to reproduce the described issue with SegmentedEditor on Android (can you confirm that you also observe it only on android). We are already preparing a fix that will be released soon.

Regarding the label, you are actually able to retrieve it quite easily even with the current version of nativescript-ui-dataform. I have [updated the playground link](https://play.nativescript.org/?template=play-js&id=0NqLMI&v=7) to illustrate how you can take note of both the id and the label if the committed value.

**Comment by pbgc** [comment]
Hi @tgpetrov  !

Thanks for you answer!

> (can you confirm that you also observe it only on android)
I only use Android .. so I currently cannot confirm on IOS. I can try to quickly build the app on IOS to check it.

> Regarding the label, you are actually able to retrieve it quite easily even with the current version of nativescript-ui-dataform. I have [updated the playground link](https://play.nativescript.org/?template=play-js&id=0NqLMI&v=7) to illustrate how you can take note of both the id and the label if the committed value.

I will try it ! Thanks!

**Comment by tgpetrov** [comment]
> I only use Android .. so I currently cannot confirm on IOS.

Don't worry, I just wanted to ensure that you need a fix for Android.

**Comment by tgpetrov** [comment]
I'd just like to sum things up, since it became a long thread:
1. When DatePicker has initial value `2016-05-16` and the value is changed, the new value is `1543687642999` instead of a string. - I never managed to reproduce this. Does any has more info or steps to reproduce?
2. When DatePicker has initial value of `null` and the value is changed, the new value is `1543687642999`. - I reproduced this, it looks odd and we need to improve this behavior
3. When using key-value pairs as `valueProvider`, `SegmentedEditor` ignores initial value on Android - I reproduced this and logged this as a [separate issue](https://github.com/telerik/nativescript-ui-feedback/issues/1031) so its progress can be tracked easier
4. When using key-value pairs as `valueProvider`, entityProperty.value returns the key and not the label - it is expected, in order to access the label you can convert the key as shown [here](https://play.nativescript.org/?template=play-js&id=0NqLMI&v=7)

**Comment by pbgc** [comment]
@tgpetrov 

>     1. When DatePicker has initial value `2016-05-16` and the value is changed, the new value is `1543687642999` instead of a string. - I never managed to reproduce this. Does any has more info or steps to reproduce?

I don't have this problem. 
Like I said above "I initialize a datepicker field with "2018-11-30" and then change it to "2018-12-01" .. the returned value is the string "2018-12-01".

>     2\. When DatePicker has initial value of `null` and the value is changed, the new value is `1543687642999`. - I reproduced this, it looks odd and we need to improve this behavior

This is the big problem I have! I don't want to initialize some DatePickers! when I create data with a form the values are null! Or when I edit a model where some Dates are initialized and other no.
This gives me an inconsistent behaviour! But only with TNS > 4.2 ... on 3.7 this works well!

>     3\. When using key-value pairs as `valueProvider`, `SegmentedEditor` ignores initial value on Android - I reproduced this and logged this as a [separate issue](https://github.com/telerik/nativescript-ui-feedback/issues/1031) so its progress can be tracked easier

Great! thanks! This really needs to be solved... its a big problem

>     4\. When using key-value pairs as `valueProvider`, entityProperty.value returns the key and not the label - it is expected, in order to access the label you can convert the key as shown [here](https://play.nativescript.org/?template=play-js&id=0NqLMI&v=7)

Thanks!! I was having a hard time with this! I used the modified behaviour all over the place and was struggling to fix it.

**Comment by marcelomiranda90** [comment]
I still have the problem on Android, the last version that works correctly is 3.7.2, I always switch to a build for Android and the latest version for iOS.

In the class step step a string with the date in the format yyyy-mm-dd and if no field change is performed and I request the update with the object, my date comes as timestamp (example 1542765600000), any change in the DatePicker and the return of my date comes as a string in the format yyyy-mm-dd.

I use this code snippet to create a timestamp, but if no change is made in the DatePicker, the split function is not recognized because the data type is number.

```
     let parts = this._carteira_editar.data.split ('-');
     mydate = new Date (parseInt (parts [0]), (parseInt (parts [1]) - 1), parseInt (parts [2]));
```

For now I keep version 3.7.2 for Android and the last version for iOS that has the problem with the decimal pt-br resolved.

**Comment by tgpetrov** [comment]
Ok guys, so the remaining issues are:
1. Issue was reported by @marcelomiranda90: When DatePicker has initial value `2016-05-16` and the value is changed, the new value is `1543687642999` instead of a string.

I managed to reproduce it with [this playground](https://play.nativescript.org/?template=play-ng&id=zRcoms). The issue will be fixed in with the next release of the plugin.

2. Issue was reported by @pbgc: When DatePicker has initial value of `null` and the value is changed, the new value is `1543687642999`. 

I managed to reproduce it with [this playground](https://play.nativescript.org/?template=play-ng&id=2P5gq1). With the next release of the plugin the behavior will be:
- if initial value is `null` or is of type `Date` - the new value will be of type `Date`
- if initial value is `""` or is of type `string` (like 2016-05-16) - the new value will be of type `string` (like 2016-05-16)
